{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e36f8907-d88c-4af0-ba62-159f64762105",
   "metadata": {},
   "source": [
    "# Variant exploration\n",
    "We've used the `nf-core/sarek` pipeline to call variants in our samples; now let's explore the data we have!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b999f12c-2d58-4bc6-9ace-27e20bc7aeb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pysam import VariantFile\n",
    "from collections import defaultdict\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a40c69b7-151f-4c02-8ae0-2784fd57542e",
   "metadata": {},
   "source": [
    "## Read in the data\n",
    "`0M_R1` was grown without EMS in 10/10\n",
    "\n",
    "`0_1M_R1` was treated with EMS in 10/10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1eee9263-65be-4224-ae8f-dcad2c4c4d43",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_0M_R1 = VariantFile('../data/nanno_sarek/variant_calling/strelka/0M_R1_vs_537_control_R2/0M_R1_vs_537_control_R2.strelka.somatic_snvs.vcf.gz') \n",
    "vcf_0_1M_R2 = VariantFile('../data/nanno_sarek/variant_calling/strelka/0_1M_R2_vs_537_control_R2/0_1M_R2_vs_537_control_R2.strelka.somatic_snvs.vcf.gz') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "314a48f3-b8e9-4a27-87a3-496473882c79",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_variants_0M_R1 = vcf_0M_R1.fetch()\n",
    "all_variants_0_1M_R2 = vcf_0_1M_R2.fetch()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cffb24b7-ab88-4229-aee9-0098b9d53f4c",
   "metadata": {},
   "source": [
    "## Filtering by support\n",
    "As mentioned in [this video](https://youtu.be/zO9WCOaq3aQ?si=GZD3t4__wenFgTxh&t=665), variants that only appear in a few reads are more likely to be sequencing errors than true variants. Therefore, we want to filter our variants. While Strelka2 doesn't provide a score in the `QUAL` column of the `vcf` file, it does provide a `PASS` value based on the Empirical Variant Scoring model (see [Supplementary Note 1](https://static-content.springer.com/esm/art%3A10.1038%2Fs41592-018-0051-x/MediaObjects/41592_2018_51_MOESM1_ESM.pdf)) if the variant is good quality, or `LowEVS` or `LowDepth` if it's not a good variant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ac15254e-ea75-49cd-8762-26fac5b8bf6c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For sample 0M_R1, 51 of 41470 variants passed, which is 0.12%.\n",
      "For sample 0_1M_R2, 556 of 46391 variants passed, which is 1.20%.\n"
     ]
    }
   ],
   "source": [
    "all_alts = {}\n",
    "all_conversions = {}\n",
    "all_passed = {}\n",
    "for sample, variants in {'0M_R1': all_variants_0M_R1, '0_1M_R2': all_variants_0_1M_R2}.items():\n",
    "    passed = []\n",
    "    alts = defaultdict(int)\n",
    "    conversions = defaultdict(int)\n",
    "    total = 0\n",
    "    for variant in variants:\n",
    "        if variant.filter.keys()[0] == 'PASS':\n",
    "            # print(variant.chrom, variant.pos, variant.ref, \">\", variant.alts[0], variant.qual, variant.filter.values()[0].name)\n",
    "            passed.append(variant)\n",
    "            alts[variant.alts[0]] += 1\n",
    "            conversions[f'{variant.ref} > {variant.alts[0]}'] += 1\n",
    "        total += 1\n",
    "    print(f'For sample {sample}, {len(passed)} of {total} variants passed, which is {(len(passed)/total)*100:.2f}%.')\n",
    "    all_alts[sample] = alts\n",
    "    all_conversions[sample] = conversions\n",
    "    all_passed[sample] = passed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f67da33-61a6-479f-8832-782f8ead0962",
   "metadata": {},
   "source": [
    "That's great! It substantially narrows down our analysis. Let's save out the filtered subset to look at in IGV:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ba5e4afa-b325-4f1a-adff-95b89e7b77b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Adapted from the Gemini suggested answer\n",
    "for sample in ['0M_R1', '0_1M_R2']:\n",
    "    with VariantFile(f'../data/nanno_sarek/variant_calling/strelka/{sample}_vs_537_control_R2/{sample}_vs_537_control_R2.strelka.somatic_snvs.vcf.gz') as vcf_in, \\\n",
    "         VariantFile(f'../data/nanno_sarek/variant_calling/strelka/{sample}_vs_537_control_R2/{sample}_vs_537_control_R2_FILTERED.strelka.somatic_snvs.vcf.gz', 'w', header=vcf_in.header) as vcf_out:\n",
    "    \n",
    "        for record in vcf_in:\n",
    "            if 'PASS' in record.filter.keys():\n",
    "                vcf_out.write(record)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbb877d0-ddf5-4c04-b54c-110d91d052ab",
   "metadata": {},
   "source": [
    "## Digging into SNVs\n",
    "Let's take a look at the landscape of variants quantitatively, beyond just looking at the viewer. First, let's look at the rates of the different transitions (which I'll calculate here as just the fraction of times the transition occurred of the variants that passed the screening -- I think this is different than the calculations done in the Arabidopsis paper linked below). Does this match with published expectations of the rates of these transitions?\n",
    "\n",
    "We'll pair together transitions that are complements of one another, as was done in [Ossowski et al, 2014, Table 2](https://pmc.ncbi.nlm.nih.gov/articles/PMC3878865/table/T2/):\n",
    "```\n",
    "A:T→G:C\n",
    "C:G→T:A\n",
    "A:T→T:A\n",
    "C:G→A:T\n",
    "A:T→C:G\n",
    "C:G→G:C\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "621da2aa-9d9d-4dbc-a4ea-8f251a3b4c98",
   "metadata": {},
   "outputs": [],
   "source": [
    "conversion_pairs = {\n",
    "    'A:T→G:C': ['A > G', 'T > C'],\n",
    "    'C:G→T:A': ['C > T', 'G > A'],\n",
    "    'A:T→T:A': ['A > T', 'T > A'],\n",
    "    'C:G→A:T': ['C > A', 'G > T'],\n",
    "    'A:T→C:G': ['A > C', 'T > G'],\n",
    "    'C:G→G:C': ['C > G', 'G > C']\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4780449a-7d22-43ab-9f3d-e908c695f1a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "conversion_probs = defaultdict(dict)\n",
    "for sample, conversions in all_conversions.items():\n",
    "    for name, pair in conversion_pairs.items():\n",
    "        pair_total = conversions[pair[0]] + conversions[pair[1]]\n",
    "        conversion_probs[sample][name] = pair_total/len(all_passed[sample])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "07c4583f-833b-410b-9044-e4e6082ae789",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Need tow rangle for stacked bars\n",
    "conversions_to_plot = {pair: np.array([v[pair] for v in conversion_probs.values()]) for pair in conversion_pairs.keys()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e09485e4-7e8f-4152-816c-3e4064d9e0f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'A:T→G:C': array([0.23529412, 0.03956835]),\n",
       " 'C:G→T:A': array([0.09803922, 0.86330935]),\n",
       " 'A:T→T:A': array([0.07843137, 0.02517986]),\n",
       " 'C:G→A:T': array([0.19607843, 0.02697842]),\n",
       " 'A:T→C:G': array([0.21568627, 0.02158273]),\n",
       " 'C:G→G:C': array([0.17647059, 0.02338129])}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "conversions_to_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "964b97ba-d18c-4d44-ae6d-02786eefca79",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjcAAAGdCAYAAADuR1K7AAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjAsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvlHJYcgAAAAlwSFlzAAAPYQAAD2EBqD+naQAAQH5JREFUeJzt3XlclOX+//H3yOaCgguSCyIuoUcyDbPUrNwz28wFNbfQjuaWSmnoOYkel7Q08yR6SpYsTUwt+5bHIsvd1AyyErEURA00N3ApEJjfH/6c08TiDA6O3L6ej8c8Hs41931dn4lHN2+u+76v22Q2m80CAAAwiHLOLgAAAMCRCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQXJ1dwM2Wn5+vX3/9VZUrV5bJZHJ2OQAAwAZms1kXLlxQ7dq1Va5c8XMzt124+fXXX+Xn5+fsMgAAQAkcO3ZMdevWLXab2y7cVK5cWdLV/zhVqlRxcjUAAMAWWVlZ8vPzs/weL85tF26unYqqUqUK4QYAgDLGlktKuKAYAAAYCuEGAAAYCuEGAAAYCuEGAAAYCuEGAAAYCuEGAAAYCuEGAAAYCuEGAAAYCuEGAAAYCuEGAAAYilPDzdatW/X444+rdu3aMplM+vjjj6+7z5YtWxQcHKzy5curQYMGWrp0aekXCgAAygynhptLly7p7rvv1ltvvWXT9ikpKXr00UfVvn17JSQkaMqUKRo3bpzWrl1bypUCAICywqkPzuzevbu6d+9u8/ZLly5VvXr1tHDhQklS06ZN9e233+r1119Xr169SqlKAABQlpSpa2527dqlrl27WrV169ZN3377ra5cuVLoPtnZ2crKyrJ6AQAA43LqzI29MjIy5Ovra9Xm6+ur3NxcnT59WrVq1Sqwz5w5czR9+vSbVaLmhzx208a6nYXFfersEgCUQRyjbw5nH6PL1MyNJJlMJqv3ZrO50PZrwsPDlZmZaXkdO3as1GsEAADOU6Zmbu644w5lZGRYtZ06dUqurq6qXr16oft4eHjIw8PjZpQHAABuAWVq5qZNmzaKj4+3avviiy/UqlUrubm5OakqAABwK3FquLl48aISExOVmJgo6eqt3omJiUpLS5N09ZTS4MGDLduPHDlSR48e1cSJE5WUlKTo6GhFRUXpxRdfdEb5AADgFuTU01LffvutOnToYHk/ceJESdKQIUMUGxur9PR0S9CRpICAAG3YsEETJkzQ4sWLVbt2bS1atIjbwAEAgIVTw83DDz9suSC4MLGxsQXaHnroIX333XelWBUAACjLytQ1NwAAANdDuAEAAIZSpm4FBwDgRpSvOtHZJeAmYOYGAAAYCuEGAAAYCuEGAAAYCuEGAAAYCuEGAAAYCuEGAAAYCreCAwBuGx03j3Z2CbeJJKeOzswNAAAwFMINAAAwFMINAAAwFMINAAAwFMINAAAwFMINAAAwFMINAAAwFMINAAAwFMINAAAwFFYoBgDcNvqG82vvZvjByeMzcwMAAAyFcAMAAAyFcAMAAAyFcAMAAAyFcAMAAAyFcAMAAAyFcAMAAAyFG/4BALeNH1LSnF0CbgJmbgAAgKEQbgAAgKEQbgAAgKEQbgAAgKEQbgAAgKEQbgAAgKEQbgAAgKEQbgAAgKEQbgAAgKEQbgAAgKEQbgAAgKEQbgAAgKEQbgAAgKEQbgAAgKEQbgAAgKEQbgAAgKEQbgAAgKEQbgAAgKEQbgAAgKEQbgAAgKEQbgAAgKEQbgAAgKEQbgAAgKEQbgAAgKEQbgAAgKEQbgAAgKEQbgAAgKG4OrsAoylfdaKzSwAA4LbGzA0AADAUwg0AADAUwg0AADAUwg0AADAUp4ebyMhIBQQEqHz58goODta2bduK3X7FihW6++67VbFiRdWqVUvPPvuszpw5c5OqBQAAtzqnhpu4uDiNHz9eU6dOVUJCgtq3b6/u3bsrLS2t0O23b9+uwYMHa9iwYfrpp5/04Ycfau/evRo+fPhNrhwAANyqnBpuFixYoGHDhmn48OFq2rSpFi5cKD8/Py1ZsqTQ7b/55hvVr19f48aNU0BAgB544AGNGDFC33777U2uHAAA3KqcFm5ycnK0b98+de3a1aq9a9eu2rlzZ6H7tG3bVsePH9eGDRtkNpt18uRJrVmzRj169ChynOzsbGVlZVm9AACAcTkt3Jw+fVp5eXny9fW1avf19VVGRkah+7Rt21YrVqxQSEiI3N3ddccdd8jb21v//ve/ixxnzpw58vLysrz8/Pwc+j0AAMCtxekXFJtMJqv3ZrO5QNs1Bw4c0Lhx4/TKK69o37592rhxo1JSUjRy5Mgi+w8PD1dmZqbldezYMYfWDwAAbi1Oe/xCjRo15OLiUmCW5tSpUwVmc66ZM2eO2rVrp5deekmS1Lx5c1WqVEnt27fXzJkzVatWrQL7eHh4yMPDw/FfAAAA3JKcNnPj7u6u4OBgxcfHW7XHx8erbdu2he5z+fJllStnXbKLi4ukqzM+AAAATj0tNXHiRC1btkzR0dFKSkrShAkTlJaWZjnNFB4ersGDB1u2f/zxx7Vu3TotWbJER44c0Y4dOzRu3Di1bt1atWvXdtbXAAAAtxCnPhU8JCREZ86c0YwZM5Senq6goCBt2LBB/v7+kqT09HSrNW+GDh2qCxcu6K233lJYWJi8vb3VsWNHzZ0711lfAQAA3GJM5tvsfE5WVpa8vLyUmZmpKlWqOLz/xSO/cnifKGj00o7OLgFAWRTh5ewKbg8RmQ7v0p7f306/WwoAAMCRCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQXJ1dAICbz2w2Kzc3V3l5ec4uBXZwc3OTi4uLs8sAbnmEG+A2k5OTo/T0dF2+fNnZpcBOJpNJdevWlaenp7NLAW5phBvgNpKfn6+UlBS5uLiodu3acnd3l8lkcnZZsIHZbNZvv/2m48ePq3HjxszgAMUg3AC3kZycHOXn58vPz08VK1Z0djmwk4+Pj1JTU3XlyhXCDVAMLigGbkPlyvG/flnELBtgG45wAADAUAg3AADAULjmBoAkqf7Ln93U8VJf7VGi/Xbu3Kn27durS5cu2rhxY+F9p6YqICCg2H6mTZumiIgIu8fPysrSa6+9pnXr1unIkSOqWLGiGjRooD59+ui5555T1apVi91/7dq1+ve//62EhATl5eWpQYMG6t27t8aMGaNq1arZXQ+Agpi5AVCmREdHa+zYsdq+fbvS0tIK3cbPz0/p6emWV1hYmJo1a2bV9uKLL9o99tmzZ3X//fcrJiZGL774onbv3q0dO3Zo2rRpSkxM1MqVK4vdf+rUqQoJCdG9996r//73v/rxxx81f/58ff/993rvvffsrgdA4Zi5AVBmXLp0SatXr9bevXuVkZGh2NhYvfLKKwW2c3Fx0R133GF57+npKVdXV6u260lJSVGdOnXk7u5uaZsyZYrS0tKUnJysOnXqWNqbNGmixx57TGazucj+9uzZo9mzZ2vhwoV64YUXLO3169dXly5ddP78eZtrA1A8Zm4AlBlxcXEKDAxUYGCgBg4cqJiYGEugSE1Nlclk0ubNmx0y1ogRI9S7d2/l5ORIurpGUFxcnAYOHGgVbP7sz3czRUREqH79+pb3K1askKenp0aNGlXovt7e3g6pGwDhBkAZEhUVpYEDB0qSHnnkEV28eFGbNm2SdPXRBIGBgQ5bv2flypU6evSoevXqpZycHP322286f/68AgMDrbYLDg6Wp6enPD091b9/f0t7jRo11LBhQ8v7n3/+WQ0aNJCbm5tD6gNQNMINgDIhOTlZe/bsUb9+/SRJrq6uCgkJUXR0tCSpTp06OnjwoFq3bm1zn71795bJZCr05ePjo/379+vTTz/Vm2++adnnr2vNfPTRR0pMTFS3bt30+++/W9rHjBljCV7S1RWGWacGuDm45gZAmRAVFaXc3FyrU0Jms1lubm46d+7cde9SKsyiRYs0c+bMQj/Lzs7WkCFDZDKZFBoaqqpVq8rb21sHDx602q5evXqSpMqVKxd73cydd96p7du368qVK8zeAKWMmRsAt7zc3FwtX75c8+fPV2JiouX1/fffy9/fXytWrChRv7Vr11aTJk0Kfc2YMUPlypXTpk2bVL16dZUrV059+/bV+++/rxMnTtg91oABA3Tx4kVFRkYW+jkXFAOOU+Jwk5OTo+TkZOXm5jqyHgAo4NNPP9W5c+c0bNgwBQUFWb169+6tqKgonThxQk2aNNGePXscMmZYWJi+/PJLq7VnZs+erTp16ui+++5TdHS09u/fr8OHD+ujjz7Srl27rJ739NZbb6lTp06W9/fdd58mTZqksLAwTZo0Sbt27dLRo0e1adMm9enTR++++65D6gZQgnBz+fJlDRs2TBUrVlSzZs0s60yMGzdOr776qsMLBICoqCh17txZXl5eBT7r1auXEhMTdfLkSSUnJ+vy5csOGbNt27YFFtWrXr269uzZo8GDB+u1115T69atdddddykiIkIhISF65513LNuePn1ahw8fttp/7ty5WrlypXbv3q1u3bqpWbNmmjhxopo3b64hQ4Y4pG4Akslc3MIMhXjhhRe0Y8cOLVy4UI888oj279+vBg0a6JNPPtG0adOUkJBQWrU6RFZWlry8vJSZmakqVao4vP/FI79yeJ8oaPTSjs4uoUz6448/lJKSooCAAJUvX97Z5cBO/PwcIKJgQEYpiMh0eJf2/P62+4Lijz/+WHFxcbr//vutrvz/29/+VuCvFAAAgJvN7tNSv/32m2rWrFmg/dKlS9zmCAAAnM7ucHPvvffqs8/+94C9a4HmnXfeUZs2bRxXGQAAQAnYfVpqzpw5euSRR3TgwAHl5ubqzTff1E8//aRdu3Zpy5YtpVEjAACAzeyeuWnbtq127Nihy5cvq2HDhvriiy/k6+urXbt2KTg4uDRqBAAAsFmJVii+6667WJMBAADckuyeuXFxcdGpU6cKtJ85c8ZqASsAAABnsDvcFLUsTnZ2ttzd3W+4IAAAgBth82mpRYsWSbp6d9SyZcvk6elp+SwvL09bt25VkyZNHF8hAACAHWwON2+88YakqzM3S5cutToF5e7urvr162vp0qWOrxAAAMAONoeblJQUSVKHDh20bt06Va1atdSKAuAEN3tZ+hIuz56RkaFZs2bps88+04kTJ1SzZk21aNFC48ePt3pQ5Z+ZzWYtW7ZM0dHR+umnn5Sfny9/f3917txZY8eOVaNGjeyqYfPmzerQoUOx28TExGjo0KGFfta1a1dt2rRJO3bs0P3332/X2ACuz+67pb7++uvSqAMAris1NVXt2rWTt7e35s2bp+bNm+vKlSv6/PPPNXr0aB08eLDAPmazWQMGDNDHH3+sKVOm6I033lDNmjWVkpKi+Ph4zZw5U7GxsXbV0bZtW6Wnp1vev/DCC8rKylJMTIylrbCHfEpSWlqadu3apTFjxigqKopwA5SCEt0Kfvz4cX3yySdKS0tTTk6O1WcLFixwSGEA8FejRo2SyWTSnj17VKlSJUt7s2bNFBoaWug+cXFxWrVqldavX68nnnjC0t6gQQN16tSpyJsk/io5OVmBgYGSrp6Kv+OOOyyfVahQQdnZ2VZtRYmJidFjjz2m559/Xq1bt9bChQutvguAG2f33VKbNm1SYGCgIiMjNX/+fH399deKiYlRdHS0EhMTS6FEAJDOnj2rjRs3avTo0YWGAW9vb0nS0KFD9fDDD1vaP/jgAwUGBloFmz+z5Zl4qampatmypVasWGFzvREREapfv75Vm9lsVkxMjAYOHKgmTZrozjvv1OrVq23uE4Bt7A434eHhCgsL048//qjy5ctr7dq1OnbsmB566CH16dOnNGoEAP3yyy8ym83XvSuzVq1aqlevnuX9oUOHLDMu14wfP16enp7y9PRU3bp1rzt2/fr1tXr1aj333HN67733bKq3Ro0aatiwoVXbl19+qcuXL6tbt26SpIEDByoqKsqm/gDYzu5wk5SUpCFDhkiSXF1d9fvvv8vT01MzZszQ3LlzHV4gAEj/W2PrejMtc+bM0fLly63a/rrP1KlTlZiYqFdeeUUXL16UJF28eFEmk6nI1+OPP67ff/9dQ4cO1cmTJ69b75gxY7Rp0yartqioKIWEhMjV9eoVAf3799fu3buVnJx83f4A2M7ua24qVaqk7OxsSVLt2rV1+PBhNWvWTJJ0+vRpx1YHAP9f48aNZTKZlJSUpKeeesqu/f56obGPj498fHxUs2ZNS1ulSpWUlJRUZD+JiYkaOnSoJk2aJF9fX7vrP3v2rD7++GNduXJFS5YssbTn5eUpOjqaPw4BB7J75ub+++/Xjh07JEk9evRQWFiYZs2apdDQ0BJd9R8ZGamAgACVL19ewcHB2rZtW7HbZ2dna+rUqfL395eHh4caNmyo6Ohou8cFULZUq1ZN3bp10+LFi3Xp0qUCn58/f77Q/fr376/k5GStX7++2P5NJpOaNGlS6KtChQoaPXq0Jk+erBkzZpSo/hUrVqhu3br6/vvvlZiYaHktXLhQ7777rnJzc0vUL4CC7J65WbBggWUaNyIiQhcvXlRcXJwaNWpkWejPVnFxcRo/frwiIyPVrl07/ec//1H37t114MABq3Pmf9a3b1+dPHlSUVFRatSokU6dOnVLHRQ6bh7t7BJuE0X/hQ3jioyMVNu2bdW6dWvNmDFDzZs3V25uruLj47VkyRIlJSUpPDxcJ06csJya6tevn9atW6d+/fopPDxc3bp1k6+vr44ePaq4uDibnolXr149LV++XD169LC51rfeeksfffSR5dRUVFSUevfuraCgIKvt/P39NXnyZH322Wd68skn7fivAaAodoebBg0aWP5dsWJFRUZGlnjwBQsWaNiwYRo+fLgkaeHChfr888+1ZMkSzZkzp8D2Gzdu1JYtW3TkyBFVq1ZNkgrcjQDAuAICAvTdd99p1qxZCgsLU3p6unx8fBQcHGw51ZOenq60tDTLPiaTSXFxcXrnnXcUExOjefPm6cqVK6pbt646depk0/IVJpPJrmAjXT1Nf/jwYUnSvn379P333+udd94psF3lypXVtWtXRUVFEW4ABzGZbV3k4TrWrVuniIgI7d+/36btc3JyVLFiRX344Yfq2bOnpf2FF15QYmKitmzZUmCfUaNG6dChQ2rVqpXee+89VapUSU888YT+9a9/qUKFCoWOk52dbblGSJKysrLk5+enzMxMValSxc5veX1JTZo6vE8U1PQgMzcl8ccffyglJcVyKhhlCz8/B7jZK3Hfrkq4AnlxsrKy5OXlZdPvb7uuuXnnnXfUp08fDRgwQLt375YkffXVV2rZsqUGDhyoNm3a2NzX6dOnlZeXV+DCPF9fX2VkZBS6z5EjR7R9+3b9+OOP+uijj7Rw4UKtWbNGo0cXfSpozpw58vLysrz8/PxsrhEAAJQ9Noeb119/XaNHj1ZKSorWr1+vjh07avbs2erbt6+eeuoppaWl6T//+Y/dBfz1Fk2z2VzkrZ75+fkymUxasWKFWrdurUcffVQLFixQbGysfv/990L3CQ8PV2ZmpuV17Ngxu2sEAABlh83X3ERFRWnp0qUKDQ3V5s2b1bFjR3311Vf65ZdfLCuD2qNGjRpycXEpMEtz6tSpIm+zrFWrlurUqWP1zJamTZvKbDbr+PHjaty4cYF9PDw85OHhYXd9AACgbLJ55ubo0aPq3LmzJOnhhx+Wm5ubZs2aVaJgI119NktwcLDi4+Ot2uPj49W2bdtC92nXrp1+/fVXy91a0tXVR8uVK2fTKqMAAMD4bA43f/zxh9UFbO7u7vLx8bmhwSdOnKhly5YpOjpaSUlJmjBhgtLS0jRy5EhJV08pDR482LL9gAEDVL16dT377LM6cOCAtm7dqpdeekmhoaFFXlAMAABuL3bdCr5s2TJ5enpKknJzcxUbG6saNWpYbTNu3Dib+wsJCdGZM2c0Y8YMpaenKygoSBs2bJC/v7+kgrd0enp6Kj4+XmPHjlWrVq1UvXp19e3bVzNnzrTnawAAAAOz+Vbw+vXrX/eZLiaTSUeOHHFIYaXFnlvJSoJbwW8ObgUvGW4lLtv4+TkAt4LfHE6+FdzmmZvU1NQbrQsAAKDU2f1sKQAAgFsZ4QYAABiK3c+WAmBMd717100d74chP5Rov507d6p9+/bq0qWLNm7cWOg2qampCggIKLafadOmKSIiwuZx69evr6NHjxb5+UMPPaTNmzcX+tnKlSs1aNAgPffcc1q6dKnNYwIoGcINgDIlOjpaY8eO1bJly5SWlqZ69eoV2MbPz0/p6emW96+//ro2btyoL7/80tJ27c5PW+3du1d5eXmSrgasXr16KTk52XJho7u7e7E1T5o0SUuWLNGCBQtUsWJFu8YGYB/CDYAy49KlS1q9erX27t2rjIwMxcbG6pVXXimwnYuLi+644w7Le09PT7m6ulq1XU9KSorq1KljCS1/XterWrVqkqSaNWtedyHT1NRU7dy5U2vXrtXXX3+tNWvWWK3fBcDxSnTNTX5+vg4dOqTt27dr69atVi8AKC1xcXEKDAxUYGCgBg4cqJiYGF1bzSI1NVUmk6nIU0P2GjFihHr37q2cnBybti9q/OjoaPXo0UNeXl4aOHCgoqKiHFIfgKLZHW6++eYbNWrUSE2bNtWDDz6ohx9+2PLq0KFDadQIAJKuPuNu4MCBkqRHHnlEFy9e1KZNmyRJbm5uCgwMdNgpn5UrV+ro0aPq1auXTQGnsPHz8/MVGxtrqblfv37atWuXfvnlF4fUCKBwdoebkSNHqlWrVvrxxx919uxZnTt3zvI6e/ZsadQIAEpOTtaePXvUr18/SZKrq6tCQkIUHR0tSapTp44OHjyo1q1b29xn7969ZTKZCn35+Pho//79+vTTT/Xmm29et6/Cxv/iiy906dIlde/eXdLVBwZ37drVUjOA0mH3NTc///yz1qxZo0aNGpVGPQBQqKioKOXm5qpOnTqWNrPZLDc3N507d05Vq1a1u89FixYV+fiW7OxsDRkyRCaTSaGhoSWqOTo6WmfPni0wm5OQkKB//etfcnFxKVG/AIpnd7i577779MsvvxBuANw0ubm5Wr58uebPn6+uXbtafdarVy+tWLFCY8aMsbvf2rVrq3bt2oV+1qtXL5UrV05ffvml5QJie5w5c0br16/XqlWr1KxZM0t7fn6+2rdvr//+97967LHH7O4XwPXZHW7Gjh2rsLAwZWRk6K677pKbm5vV582bN3dYcQAgSZ9++qnOnTunYcOGycvL+tlAvXv3VlRUlHr27KlOnTpp+fLldp2aKkpYWJiaNGlic7A5ceKE1fjvvfeeqlevrj59+qhcOesrAB577DFFRUURboBSYne46dWrlyRZTdOaTCaZzWaZTCbLOhAA4ChRUVHq3LlzgWAjXT0mzZ49WydPnlRycrIuX77skDHbtm1r1/ZXrlyxGj86Olo9e/YsEGyu1RwSEqKTJ0/K19fXIfUC+B+bnwp+TXErdEqSv7//DRVU2ngquDHwVPCS4anSZRs/PwfgqeA3R1l5Kvg1t3p4AQAAt7cSrVB8+PBhLVy4UElJSTKZTGratKleeOEFNWzY0NH1AQAA2MXudW4+//xz/e1vf9OePXvUvHlzBQUFaffu3WrWrJni4+NLo0YAAACb2T1z8/LLL2vChAl69dVXC7RPnjxZXbp0cVhxAAAA9rJ75iYpKUnDhg0r0B4aGqoDBw44pCgAAICSsjvc+Pj4KDExsUB7YmKiatas6YiaAAAASszu01LPPfec/v73v+vIkSNq27atTCaTtm/frrlz5yosLKw0agQAALCZ3eHmn//8pypXrqz58+crPDxc0tUlzCMiIjRu3DiHFwgAAGAPu8ONyWTShAkTNGHCBF24cEGSVLlyZYcXBgAAUBIlWufmGkINAAC41dgUbu655x5t2rRJVatWVcuWLWUymYrc9rvvvnNYcQBunpv96JCSPkIjIyNDs2bN0meffaYTJ06oZs2aatGihcaPH69OnToVuo/ZbNayZcsUHR2tn376Sfn5+fL391fnzp01duxYNWrU6Ea+imbPnq1//vOfmjVrll5++eUit4uNjdWzzz5bbF9ff/21Hn744RuqB7jd2RRunnzySXl4eFj+XVy4AYDSkpqaqnbt2snb21vz5s1T8+bNdeXKFX3++ecaPXq0Dh48WGAfs9msAQMG6OOPP9aUKVP0xhtvqGbNmkpJSVF8fLxmzpyp2NjYG6orJiZGkyZNUnR0dLHhJiQkRI888ojl/dNPP62goCDNmDHD0mbrU8gBFM2mcDNt2jTLvyMiIkqrFgAo1qhRo2QymbRnzx5VqlTJ0t6sWTOFhoYWuk9cXJxWrVql9evX64knnrC0N2jQQJ06dZKtzw5OTk5WYGBggfYtW7bo999/14wZM7R8+XJt3bpVDz74YKF9VKhQQRUqVLC8d3d3V8WKFXXHHXfYVAMA29i9zk2DBg105syZAu3nz59XgwYNHFIUAPzV2bNntXHjRo0ePdoq2Fzj7e0tSRo6dKjVaZ0PPvhAgYGBVsHmz2yZiU5NTVXLli21YsWKAp9FRUWpf//+cnNzU//+/RUVFWX1eWxsLLPdwE1md7hJTU1VXl5egfbs7GwdP37cIUUBwF/98ssvMpvNatKkSbHb1apVS/Xq1bO8P3ToUIEZl/Hjx8vT01Oenp6qW7fudceuX7++Vq9ereeee07vvfeepT0rK0tr167VwIEDJUkDBw7UmjVrlJWVZdnGy8ur0BkfAKXH5rulPvnkE8u/P//8c3l5eVne5+XladOmTQoICHBsdQDw/107fXS9WZA5c+YUaPvrPlOnTtWYMWO0bt06zZ49W5J08eJFm+4AHTp0qLp27SpfX1+tXLlSDRo00N133y1JatGihRo0aKBVq1bp73//uySpZ8+e6tmz5/W/IACHsTncPPXUU5KuHiSGDBli9Zmbm5vq16+v+fPnO7Q4ALimcePGMplMSkpKshyPbN3vrxca+/j4yMfHx+qRMZUqVVJSUtF3cCUmJmro0KGaNGmSfH19Jcly95Wr6/8Opfn5+YqKirKEGwA3n83hJj8/X5IUEBCgvXv3qkaNGqVWFAD8VbVq1dStWzctXrxY48aNK3Ddzfnz5y3X3fxZ//79NWDAAK1fv15PPvlkkf2bTKYiT3kdPXpUo0eP1uTJkzV9+nRJ0g8//KBvv/1WmzdvtrrD6fz583rwwQf1448/KigoqATfFMCNsvuam5SUFIINAKeIjIxUXl6eWrdurbVr1+rnn39WUlKSFi1apDZt2kiSwsPDNXjwYMs+/fr1U+/evdWvXz/NmDFDu3fvVmpqqrZs2aK4uDi5uLhcd9x69epp+fLllmAjXb2QuHXr1nrwwQcVFBRkeT3wwANq06aN5cLijz766LrXCQFwLJtmbhYtWqS///3vKl++vBYtWlTstjxfCkBpCQgI0HfffadZs2YpLCxM6enp8vHxUXBwsJYsWSJJSk9PV1pammUfk8mkuLg4vfPOO4qJidG8efN05coV1a1bV506ddKCBQuuO67JZFKPHj0s73NycvT+++9r8uTJhW7fq1cvzZkzR3PnzlVmZqaSk5Nv8JsDsIfJbMMiDwEBAfr2229VvXr1Yi8aNplMOnLkiEMLdLSsrCx5eXkpMzNTVapUcXj/N3uV19tVSVe3vd398ccfSklJUUBAgMqXL+/scmAnfn4OEOF1/W1w4yIyHd6lPb+/bZq5SUlJKfTfAAAAtxq7r7n5q7y8PCUmJurcuXOOqAcAAOCG2B1uxo8fb7lQLi8vTw8++KDuuece+fn5afPmzY6uDwAAwC52h5s1a9ZYFqz6v//7P6WmpurgwYMaP368pk6d6vACAQAA7GF3uDl9+rTlIW8bNmxQnz59dOedd2rYsGH64YcfHF4gAACAPewON76+vjpw4IDy8vK0ceNGde7cWZJ0+fJlm9aLAAAAKE02r1B8zbPPPqu+ffuqVq1aMplM6tKliyRp9+7dLFQFAACczu5wExERoaCgIB07dkx9+vSRh4eHJMnFxUUvv/yywwsEAACwh93hRpJ69+5doO2vD9MEAABwhhKFm02bNmnTpk06deqU5YGa10RHRzukMAAAgJKwO9xMnz5dM2bMUKtWrSzX3QAo+xaP/Oqmjjd6accS7bdz5061b99eXbp00caNGwvdJjU1tdhHxUjStGnTFBERYff4WVlZmjt3rtauXavU1FR5e3srKChIo0aNUs+ePYs9Jq5du1aLFy9WQkKCsrOz5efnp3bt2mns2LFq2bKl3bUAKJzd4Wbp0qWKjY3VoEGDSqMeAChWdHS0xo4dq2XLliktLU316tUrsI2fn5/S09Mt719//XVt3LhRX375paXN09PT7rHPnz+vBx54QJmZmZo5c6buvfdeubq6asuWLZo0aZI6duwob2/vQvedPHmy5s+fr3Hjxmn69OmqW7eu0tLStH37dk2ZMkX//e9/7a4HQOHsDjc5OTlq27ZtadQCAMW6dOmSVq9erb179yojI0OxsbF65ZVXCmzn4uJiWY9LuhpkXF1drdquJyUlRXXq1JG7u7ulbcqUKUpNTdWhQ4dUu3ZtS/udd96p/v37F/kwy2+++Ubz5s3Tm2++qXHjxlnaAwIC9NBDD8mG5xcDsIPd69wMHz5cK1euLI1aAKBYcXFxCgwMVGBgoAYOHKiYmBhLMEhNTZXJZHLYY2BGjBih3r17KycnR5KUn5+vVatW6ZlnnrEKNtdcC1DS1btK69evb/nsgw8+kKenp0aNGlXoWJzeBxzL7pmbP/74Q2+//ba+/PJLNW/eXG5ublafL1iwwGHFAcCfRUVFaeDAgZKkRx55RBcvXtSmTZvUuXNnubm5KTAwUBUrVnTIWCtXrlSnTp3Uq1cvrV27VufPn9e5c+dsWs+rRo0aatiwoeX9oUOH1KBBA0v4ka4eK/8863TixAl5eXk5pHbgdmd3uNm/f79atGghSfrxxx+tPuOvDwClJTk5WXv27NG6deskSa6urgoJCVF0dLQ6d+6sOnXq6ODBg3b12bt3b61du7bYbfbv368333xTgwcPlmTbcW7MmDEaM2aMVdtf9wsNDdUTTzyh3bt3a+DAgZyaAhzI7nDz9ddfl0YdAFCsqKgo5ebmqk6dOpY2s9ksNzc3nTt3TlWrVrW7z0WLFmnmzJmFfpadna0hQ4bIZDIpNDRUVatWVdWqVZWUlGT3OI0bN9b27dt15coVy2y3t7e3vL29dfz4cbv7A1A8u6+5AYCbLTc3V8uXL9f8+fOVmJhoeX3//ffy9/fXihUrStRv7dq11aRJk0JfM2bMULly5bRp0yZVr15d5cqVU0hIiFasWKFff/21QF+XLl1Sbm5uoeP0799fFy9eVGRkZInqBGCfEi3it3fvXn344YdKS0uzXGx3zbUpYwBwlE8//VTnzp3TsGHDClyX0rt3b0VFRalnz57q1KmTli9frtatW9/wmGFhYWrSpImqVatmaZs9e7Y2b96s++67T7NmzVKrVq3k5uambdu2ac6cOdq7d6+8vb311ltv6aOPPtKmTZskSW3atFFYWJjCwsJ09OhRPf3005bb1aOiomQymVSuHH9rAo5i9/9Nq1atUrt27XTgwAF99NFHunLlig4cOKCvvvqKi+EAlIqoqCh17ty50GNMr169lJiYqJMnTyo5OVmXL192yJht27a1CjaSVLVqVX3zzTcaOHCgZs6cqZYtW6p9+/b64IMP9Nprr1nqO336tA4fPmy17+uvv66VK1cqISFBjz32mBo3bqw+ffooPz9fu3btUpUqVRxSNwDJZLbzKrbmzZtrxIgRGj16tCpXrqzvv/9eAQEBGjFihGrVqqXp06eXVq0OkZWVJS8vL2VmZpbKwSSpSVOH94mCmh60/7oHXL3bMSUlRQEBAUWuyYJbFz8/B4jgj/CbIiLT4V3a8/vb7pmbw4cPq0ePHpIkDw8PXbp0SSaTSRMmTNDbb79dsooBAAAcxO5wU61aNV24cEGSVKdOHcvt4OfPn3fYdDAAAEBJ2X1Bcfv27RUfH6+77rpLffv21QsvvKCvvvpK8fHx6tSpU2nUCAAAYDO7w81bb72lP/74Q5IUHh4uNzc3bd++XU8//bT++c9/2l1AZGSkXnvtNaWnp6tZs2ZauHCh2rdvf939duzYoYceekhBQUFKTEy0e9zS0je8RDegwU4/OLsAAMAty67TUrm5ufq///s/yy2L5cqV06RJk/TJJ59owYIFdi+iFRcXp/Hjx2vq1KlKSEhQ+/bt1b17d6WlpRW7X2ZmpgYPHsxMEQAAKMCucOPq6qrnn39e2dnZDhl8wYIFGjZsmIYPH66mTZtq4cKF8vPz05IlS4rdb8SIERowYIDatGnjkDoAAIBx2H1B8X333aeEhIQbHjgnJ0f79u1T165drdq7du2qnTt3FrlfTEyMDh8+rGnTptk0TnZ2trKysqxeAADAuOy+QGTUqFEKCwvT8ePHFRwcrEqVKll93rx5c5v6OX36tPLy8uTr62vV7uvrq4yMjEL3+fnnn/Xyyy9r27ZtVk/XLc6cOXNu+bV3AACA49gcbkJDQ7Vw4UKFhIRIksaNG2f5zGQyyWw2y2QyKS8vz64C/vqk3Gv9/FVeXp4GDBig6dOn684777S5//DwcE2cONHyPisrS35+fnbVCAAAyg6bw827776rV199VSkpKQ4ZuEaNGnJxcSkwS3Pq1KkCszmSdOHCBX377bdKSEjQmDFjJEn5+fkym81ydXXVF198oY4dOxbYz8PDQx4eHg6pGQAA3PpsDjfXntLg7+/vkIHd3d0VHBys+Ph49ezZ09IeHx+vJ598ssD2VapU0Q8/WN8AHBkZqa+++kpr1qxRQECAQ+oCblfzQx67qeOFxX1aov0yMjI0a9YsffbZZzpx4oRq1qypFi1aaPz48UXeQWk2m7Vs2TJFR0frp59+Un5+vvz9/dW5c2eNHTtWjRo1KvH3SEhI0KuvvqqtW7fq7NmzuuOOO3TXXXdpxIgReuyxxwqdib4mKytLc+fO1dq1a5Wamipvb28FBQVp1KhR6tmzZ7H7AiiaXdfcOPp/tIkTJ2rQoEFq1aqV2rRpo7fffltpaWkaOXKkpKunlE6cOKHly5erXLlyCgoKstq/Zs2aKl++fIF2AMaUmpqqdu3aydvbW/PmzVPz5s115coVff755xo9erQOHjxYYB+z2awBAwbo448/1pQpU/TGG2+oZs2aSklJUXx8vGbOnKnY2NgS1bN+/Xr17dtXnTt31rvvvquGDRvqzJkz2r9/v/7xj3+offv28vb2LnTf8+fP64EHHlBmZqZmzpype++9V66urtqyZYsmTZqkjh07FrkvgOLZFW7uvPPO6wacs2fP2txfSEiIzpw5oxkzZig9PV1BQUHasGGDZXYoPT39umveALh9jBo1SiaTSXv27LG6maFZs2YKDQ0tdJ+4uDitWrVK69ev1xNPPGFpb9CggTp16iRbnx2cnJyswMBAy/tLly5p2LBh6tGjh9atW2dpb9iwoVq3bq3hw4cX2/eUKVOUmpqqQ4cOqXbt2pb2O++8U/379+fBmMANsCvcTJ8+XV5ejn2i6qhRozRq1KhCP7veX1MRERGKiIhwaD0Abk1nz57Vxo0bNWvWrAJ3aUqyzHIMHTpUqamp2rx5syTpgw8+UGBgoFWw+TNbZqRTU1PVsmVLvfPOO3rmmWckSV988YXOnDmjSZMmFbnfn/uuX7++hg4dqoiICOXn52vVqlV65plnrILNNZ6entetCUDR7Ao3/fr1U82aNUurFgAo0i+//CKz2awmTZoUu12tWrWUn59veX/o0CGrGRdJGj9+vJYtWybpaig6fvx4sX3Wr19fq1evVt++fZWfn69Bgwbp0KFDkmTV9969e9WhQwfL+1WrVumxx65ey9SwYUPVqFFD0tWlMM6dO3fd7wKgZGwON1zYBsCZrp3iud6xaM6cOQXa/rrP1KlTNWbMGK1bt06zZ8+WJF28eFGVK1e+bh1Dhw4tsPjoNc2bN7c8665x48bKzc21fLZp0ya7vwuAkrH7bikAcIbGjRvLZDIpKSlJTz31lF37/fVCYx8fH/n4+FjNRFeqVElJSUlF9pOYmKihQ4dq0qRJ8vX1VePGjSVdvRbn/vvvl3R16Qlb7rzy8fFR1apVix0PQMnZ/PiF/Px8TkkBcJpq1aqpW7duWrx4sS5dulTg8/Pnzxe6X//+/ZWcnKz169cX27/JZFKTJk0KfVWoUEGjR4/W5MmTNWPGDElXHxVTrVo1zZ071+7vUq5cOYWEhGjFihX69ddfC3x+6dIlq1kfAPax+9lSAOAskZGRysvLU+vWrbV27Vr9/PPPSkpK0qJFiywP0g0PD9fgwYMt+/Tr10+9e/dWv379NGPGDO3evVupqanasmWL4uLi5OLict1x69Wrp+XLl1s9ysXT01PLli3TZ599ph49eujzzz/XkSNHtH//fs2bN0+SrPru1KmT3nrrLcv72bNny8/PT/fdd5+WL1+uAwcO6Oeff1Z0dLRatGihixcv3vB/L+B2ZfezpQDAWQICAvTdd99p1qxZCgsLU3p6unx8fBQcHKwlS5ZIKriEhMlkUlxcnN555x3FxMRo3rx5unLliurWratOnTppwYIF1x3XZDKpR48eBdp79uypnTt3au7cuRo8eLDOnj0rLy8vtWrVyupiYkk6fPiwTp8+bXlftWpVffPNN3r11Vc1c+ZMHT16VFWrVtVdd92l1157zeF3pgK3E5P5NruYJisrS15eXsrMzFSVKlUc3v9d797l8D5R0A9Dfrj+Rijgjz/+UEpKigICAlhHpQzi5+cAEYTGmyIi0+Fd2vP7m9NSAADAUAg3AADAUAg3AADAUAg3AADAUAg3AADAUAg3wG3oz89eQtlxm93cCpQY69wAtxF3d3eVK1dOv/76q3x8fOTu7s7zjcoIs9ms3377TSaTSW5ubs4uB7ilEW6A20i5cuUUEBCg9PT0Qpf9x63NZDKpbt26Nq2qDNzOCDfAbcbd3V316tVTbm6u8vLynF0O7ODm5kawAWxAuAFuQ9dObXB6A4AREW4c7IeUtOtvBAAASg13SwEAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAENxdXYBRlP/j5XOLuG2kOrsAgAAtyxmbgAAgKEQbgAAgKEQbgAAgKEQbgAAgKEQbgAAgKEQbgAAgKEQbgAAgKEQbgAAgKE4PdxERkYqICBA5cuXV3BwsLZt21bktuvWrVOXLl3k4+OjKlWqqE2bNvr8889vYrUAAOBW59RwExcXp/Hjx2vq1KlKSEhQ+/bt1b17d6WlpRW6/datW9WlSxdt2LBB+/btU4cOHfT4448rISHhJlcOAABuVSaz2Wx21uD33Xef7rnnHi1ZssTS1rRpUz311FOaM2eOTX00a9ZMISEheuWVV2zaPisrS15eXsrMzFSVKlVKVHdx6r/8mcP7REGpr/ZwdgkAyqIIL2dXcHuIyHR4l/b8/nbazE1OTo727dunrl27WrV37dpVO3futKmP/Px8XbhwQdWqVStym+zsbGVlZVm9AACAcTkt3Jw+fVp5eXny9fW1avf19VVGRoZNfcyfP1+XLl1S3759i9xmzpw58vLysrz8/PxuqG4AAHBrc/oFxSaTyeq92Wwu0FaYDz74QBEREYqLi1PNmjWL3C48PFyZmZmW17Fjx264ZgAAcOtyddbANWrUkIuLS4FZmlOnThWYzfmruLg4DRs2TB9++KE6d+5c7LYeHh7y8PC44XoBAEDZ4LSZG3d3dwUHBys+Pt6qPT4+Xm3bti1yvw8++EBDhw7VypUr1aMHF5UCAABrTpu5kaSJEydq0KBBatWqldq0aaO3335baWlpGjlypKSrp5ROnDih5cuXS7oabAYPHqw333xT999/v2XWp0KFCvLy4gp4AADg5HATEhKiM2fOaMaMGUpPT1dQUJA2bNggf39/SVJ6errVmjf/+c9/lJubq9GjR2v06NGW9iFDhig2NvZmlw8AAG5BTl3nxhlY58YYWOcGQImwzs3NcbuucwMAAFAaCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQCDcAAMBQXJ1dAAAAN0v9P1Y6u4TbQqqTx2fmBgAAGArhBgAAGArhBgAAGArhBgAAGArhBgAAGArhBgAAGArhBgAAGArhBgAAGArhBgAAGIrTw01kZKQCAgJUvnx5BQcHa9u2bcVuv2XLFgUHB6t8+fJq0KCBli5depMqBQAAZYFTw01cXJzGjx+vqVOnKiEhQe3bt1f37t2VlpZW6PYpKSl69NFH1b59eyUkJGjKlCkaN26c1q5de5MrBwAAtyqnhpsFCxZo2LBhGj58uJo2baqFCxfKz89PS5YsKXT7pUuXql69elq4cKGaNm2q4cOHKzQ0VK+//vpNrhwAANyqnPbgzJycHO3bt08vv/yyVXvXrl21c+fOQvfZtWuXunbtatXWrVs3RUVF6cqVK3JzcyuwT3Z2trKzsy3vMzMzJUlZWVk3+hUKlZ99uVT6hbXS+vkBMDaO0TdHaRyjr/VpNpuvu63Tws3p06eVl5cnX19fq3ZfX19lZGQUuk9GRkah2+fm5ur06dOqVatWgX3mzJmj6dOnF2j38/O7gerhbF4LnV0BAKAopXmMvnDhgry8vIrdxmnh5hqTyWT13mw2F2i73vaFtV8THh6uiRMnWt7n5+fr7Nmzql69erHj3C6ysrLk5+enY8eOqUqVKs4uBwDwJxyj/8dsNuvChQuqXbv2dbd1WripUaOGXFxcCszSnDp1qsDszDV33HFHodu7urqqevXqhe7j4eEhDw8PqzZvb++SF25QVapUue3/xwGAWxXH6KuuN2NzjdMuKHZ3d1dwcLDi4+Ot2uPj49W2bdtC92nTpk2B7b/44gu1atWq0OttAADA7cepd0tNnDhRy5YtU3R0tJKSkjRhwgSlpaVp5MiRkq6eUho8eLBl+5EjR+ro0aOaOHGikpKSFB0draioKL344ovO+goAAOAW49RrbkJCQnTmzBnNmDFD6enpCgoK0oYNG+Tv7y9JSk9Pt1rzJiAgQBs2bNCECRO0ePFi1a5dW4sWLVKvXr2c9RXKPA8PD02bNq3AqTsAgPNxjC4Zk9mWe6oAAADKCKc/fgEAAMCRCDcAAMBQCDcAAMBQCDcAAMBQCDdlUGRkpAICAlS+fHkFBwdr27Ztls8efvhhmUwmvfrqqwX2e/TRR2UymRQREWHTONf6MplMcnd3V8OGDRUeHm71rC5JmjVrltq2bauKFSuyQCIAQyvu+FucdevWqVu3bqpRo4ZMJpMSExMLbFO/fn2ZTCatWrWqwGfNmjWTyWRSbGysTeNd68tkMqlChQpq0qSJXnvtNavnMn3//ffq37+//Pz8VKFCBTVt2lRvvvmmTf3f6gg3ZUxcXJzGjx+vqVOnKiEhQe3bt1f37t2tbpn38/NTTEyM1X6//vqrvvrqq0Kfv1Wc5557Tunp6frll180b948LV68uEA4ysnJUZ8+ffT888+X+HsBwK3OluNvUS5duqR27doV+ofnnxV2/P7mm2+UkZGhSpUq2VXvtWVWkpKS9OKLL2rKlCl6++23LZ/v27dPPj4+ev/99/XTTz9p6tSpCg8P11tvvWXXOLckM8qU1q1bm0eOHGnV1qRJE/PLL79sNpvN5oceesj8/PPPm6tXr27evn27ZZtZs2aZH3/8cfPdd99tnjZtmk1jPfTQQ+YXXnjBqu3pp58233PPPYVuHxMTY/by8rL5uwBAWXK9468tUlJSzJLMCQkJBT7z9/c3v/zyy2YPDw9zWlqapf25554zjx071uzl5WWOiYmxaRx/f3/zG2+8YdV2zz33mJ9++uli9xs1apS5Q4cONo1xK2PmpgzJycnRvn371LVrV6v2rl27aufOnZb37u7ueuaZZ6zSf2xsrEJDQ29o/O+//147duzgURcAbju2Hn9vlK+vr7p166Z3331XknT58mXFxcXd0PHbbDZr8+bNSkpKuu7xOzMzU9WqVSvxWLcKwk0Zcvr0aeXl5RV4sKivr2+BB4oOGzZMq1ev1qVLl7R161ZlZmaqR48edo8ZGRkpT09PeXh4qEWLFvrtt9/00ksv3dD3AICyxp7j740KDQ1VbGyszGaz1qxZo4YNG6pFixZ29zN58mTL8btDhw4ym80aN25ckdvv2rVLq1ev1ogRI26g+lsD4aYMMplMVu/NZnOBtubNm6tx48Zas2aNoqOjNWjQoBLNuDzzzDNKTEzUrl271LdvX4WGhvK4CwC3LVuOvzeqR48eunjxorZu3aro6OgSz9q89NJLSkxM1JYtW9ShQwdNnTq1yAdT//TTT3ryySf1yiuvqEuXLjdS/i3Bqc+Wgn1q1KghFxeXAn8lnDp1qsBfE9LV9L948WIdOHBAe/bsKdGYXl5eatSokSTp/fffV7NmzRQVFaVhw4aVqD8AKIvsPf7eCFdXVw0aNEjTpk3T7t279dFHH5Wonxo1aqhRo0Zq1KiR1q5dq0aNGun+++9X586drbY7cOCAOnbsqOeee07/+Mc/HPEVnI6ZmzLE3d1dwcHBio+Pt2qPj48vNI0PGDBAP/zwg4KCgvS3v/3thsd3c3PTlClT9I9//EOXL1++4f4AoKyw9/h7o0JDQ7VlyxY9+eSTqlq16g33V7VqVY0dO1Yvvvii1e3gP/30kzp06KAhQ4Zo1qxZNzzOrYJwU8ZMnDhRy5YtU3R0tJKSkjRhwgSlpaVp5MiRBbatWrWq0tPTtWnTJoeNP2DAAJlMJkVGRlra0tLSlJiYqLS0NOXl5SkxMVGJiYm6ePGiw8YFAGez5/j7V2fPnlViYqIOHDggSUpOTlZiYmKR1+s0bdpUp0+fLnBb+I0YPXq0kpOTtXbtWkn/CzZdunTRxIkTlZGRoYyMDP32228OG9NZOC1VxoSEhOjMmTOW9QuCgoK0YcMG+fv7F7q9oxfVc3d315gxYzRv3jyNHDlSnp6eeuWVVyxX9ktSy5YtJUlff/21Hn74YYeODwDOYu/x988++eQTPfvss5b3/fr1kyRNmzatyIVVq1ev7pC6r/Hx8dGgQYMUERGhp59+Wh9++KF+++03rVixQitWrLBs5+/vr9TUVIeOfbOZzH+enwIAACjjOC0FAAAMhXBzm9q2bZs8PT2LfAEAbHOzj6crVqwocqxmzZo5fLyyiNNSt6nff/9dJ06cKPLza7d/AwCKd7OPpxcuXNDJkycL/czNzc2ma4CMjnADAAAMhdNSAADAUAg3AADAUAg3AADAUAg3AADAUAg3AADAUAg3AADAUAg3AADAUAg3AADAUP4fPLemuqzeeyQAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig, ax = plt.subplots()\n",
    "\n",
    "width = 0.25\n",
    "bottom = np.zeros(2)\n",
    "\n",
    "for name, probs in conversions_to_plot.items():\n",
    "    ax.bar(conversion_probs.keys(), probs, width, bottom=bottom, label=f'{name}')\n",
    "    bottom += probs\n",
    "\n",
    "ax.set_ylabel('Transition Rate')\n",
    "ax.legend(loc='center')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8749be8-1c43-45d9-b784-a0b8fea982bd",
   "metadata": {},
   "source": [
    "This is quite cool -- EMS is [known to induce mainly C:G→T:A mutations](https://www.frontiersin.org/journals/plant-science/articles/10.3389/fpls.2022.1052569/full), and we see an enormous rate increase of that kind of mutation in the EMS sample vs. the one that was allowed to evolve on its own."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bffb0454-d407-4b8d-a248-22ef61c5cf83",
   "metadata": {},
   "source": [
    "## Indels\n",
    "Thus far we've only looked at the SNVs in our data; what about the indels?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c16563fe-7eef-479d-bb90-970c97d7c31e",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_0M_R1_indels = VariantFile('../data/nanno_sarek/variant_calling/strelka/0M_R1_vs_537_control_R2/0M_R1_vs_537_control_R2.strelka.somatic_indels.vcf.gz') \n",
    "vcf_0_1M_R2_indels = VariantFile('../data/nanno_sarek/variant_calling/strelka/0_1M_R2_vs_537_control_R2/0_1M_R2_vs_537_control_R2.strelka.somatic_indels.vcf.gz') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "50354f3d-afb7-4cef-9525-410e06a76390",
   "metadata": {},
   "outputs": [],
   "source": [
    "indels_variants_0M_R1 = vcf_0M_R1_indels.fetch()\n",
    "indels_variants_0_1M_R2 = vcf_0_1M_R2_indels.fetch()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e0803db4-dee0-4a5a-98eb-cae15421321a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Indel variants for sample 0M_R1:\n",
      "-------------------------------------------\n",
      "scf_2 453749 AC > A None PASS\n",
      "scf_9 1044578 C > CAATA None PASS\n",
      "scf_11 803969 T > TCGACGG None PASS\n",
      "scf_14 38344 CA > C None PASS\n",
      "scf_17 66764 C > CCCTT None PASS\n",
      "scf_19 567213 A > AAT None PASS\n",
      "scf_19 572560 CT > C None PASS\n",
      "scf_21 107074 CCT > C None PASS\n",
      "scf_21 196997 GTGTTCA > G None PASS\n",
      "scf_21 197209 GT > G None PASS\n",
      "scf_21 197734 GC > G None PASS\n",
      "scf_29 211683 C > CAT None PASS\n",
      "scf_35 48363 GTA > G None PASS\n",
      "scf_36 103574 TC > T None PASS\n",
      "\n",
      "For sample 0M_R1, 14 of 422 variants passed, which is 3.32%.\n",
      "\n",
      "Indel variants for sample 0_1M_R2:\n",
      "-------------------------------------------\n",
      "scf_6 712028 GCGC > G None PASS\n",
      "scf_14 627792 C > CCCCTCTCT None PASS\n",
      "scf_16 149718 A > AAT None PASS\n",
      "scf_19 707589 T > TC None PASS\n",
      "scf_20 158358 A > AG None PASS\n",
      "scf_24 819444 TG > T None PASS\n",
      "scf_24 1002851 TCTGGAA > T None PASS\n",
      "scf_29 211683 CAT > C None PASS\n",
      "scf_32 550041 C > CG None PASS\n",
      "\n",
      "For sample 0_1M_R2, 9 of 428 variants passed, which is 2.10%.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for sample, variants in {'0M_R1': indels_variants_0M_R1, '0_1M_R2': indels_variants_0_1M_R2}.items():\n",
    "    print(f'Indel variants for sample {sample}:')\n",
    "    print('-------------------------------------------')\n",
    "    passed = []\n",
    "    total = 0\n",
    "    for variant in variants:\n",
    "        if variant.filter.keys()[0] == 'PASS':\n",
    "            print(variant.chrom, variant.pos, variant.ref, \">\", variant.alts[0], variant.qual, variant.filter.values()[0].name)\n",
    "            passed.append(variant)\n",
    "        total += 1\n",
    "    print(f'\\nFor sample {sample}, {len(passed)} of {total} variants passed, which is {(len(passed)/total)*100:.2f}%.\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e29ae82c-7f03-4529-ae10-c7be319bb38f",
   "metadata": {},
   "source": [
    "There are way fewer indel mutants than point mutations, and a similar number between the normal and EMS mutants, which makes sense, as EMS induces point mutations. Write them out:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "dee6db9e-06ca-4d24-a3e9-8e66d7e4568e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Adapted from the Gemini suggested answer\n",
    "for sample in ['0M_R1', '0_1M_R2']:\n",
    "    with VariantFile(f'../data/nanno_sarek/variant_calling/strelka/{sample}_vs_537_control_R2/{sample}_vs_537_control_R2.strelka.somatic_indels.vcf.gz') as vcf_in, \\\n",
    "         VariantFile(f'../data/nanno_sarek/variant_calling/strelka/{sample}_vs_537_control_R2/{sample}_vs_537_control_R2_FILTERED.strelka.somatic_indels.vcf.gz', 'w', header=vcf_in.header) as vcf_out:\n",
    "    \n",
    "        for record in vcf_in:\n",
    "            if 'PASS' in record.filter.keys():\n",
    "                vcf_out.write(record)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "145d648b-87ce-489c-943a-3e2e58d68dbd",
   "metadata": {},
   "source": [
    "## Investigating mutations\n",
    "We now want to narrow in on what genes these mutations may be impacting. Ideally, we want to see what genes are affected by the SNPs and indels. A useful tool for this is the [BEDtools](https://bedtools.readthedocs.io/en/latest/) module; specifically, we can [get the intersection of a SNP file with a genome annotation](https://www.biostars.org/p/379113/#379121). The bulk of this work will take place on the command line.\n",
    "\n",
    "First, we need to convert the `gff3` annotation file to a `gtf` file with `gffread`. From within the `data/genome` directory:\n",
    "\n",
    "```\n",
    "gffread n_salina.working_models.gff3 -T -o n_salina.working_models.gtf\n",
    "```\n",
    "(Note: I provided the full path to where I'd cloned `gffread` rather than adding to `PATH`)\n",
    "\n",
    "Next, use BEDtools to get the intersections. I'm going to start with the non-EMS sample SNV's to get a handle on how the results will look:\n",
    "```\n",
    "bedtools intersect -wa -wb -a ../nanno_sarek/variant_calling/strelka/0M_R1_vs_537_control_R2/0M_R1_vs_537_control_R2_FILTERED.strelka.somatic_snvs.vcf.gz -b n_salina.working_models.gtf > ../nanno_sarek/variant_calling/strelka/0M_R1_vs_537_control_R2/0M_R1_vs_537_control_R2_FILTERED_SNVs_overlap_with_annotations.txt\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2916c402-c2b0-4452-bcb2-51e045cde303",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>17</th>\n",
       "      <th>18</th>\n",
       "      <th>19</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>scf_1</td>\n",
       "      <td>167764</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>SOMATIC;QSS=48;TQSS=1;NT=ref;QSS_NT=48;TQSS_NT...</td>\n",
       "      <td>DP:FDP:SDP:SUBDP:AU:CU:GU:TU</td>\n",
       "      <td>163:0:0:0:3,3:4,4:3,3:153,154</td>\n",
       "      <td>200:2:0:0:4,5:8,8:9,9:177,178</td>\n",
       "      <td>scf_1</td>\n",
       "      <td>MSU_v1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>167216</td>\n",
       "      <td>168580</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"N_sal.S000450.1\"; gene_id \"N_sa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>scf_1</td>\n",
       "      <td>167764</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>SOMATIC;QSS=48;TQSS=1;NT=ref;QSS_NT=48;TQSS_NT...</td>\n",
       "      <td>DP:FDP:SDP:SUBDP:AU:CU:GU:TU</td>\n",
       "      <td>163:0:0:0:3,3:4,4:3,3:153,154</td>\n",
       "      <td>200:2:0:0:4,5:8,8:9,9:177,178</td>\n",
       "      <td>scf_1</td>\n",
       "      <td>MSU_v1</td>\n",
       "      <td>exon</td>\n",
       "      <td>167216</td>\n",
       "      <td>168580</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"N_sal.S000450.1\"; gene_id \"N_sa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>scf_1</td>\n",
       "      <td>167764</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>SOMATIC;QSS=48;TQSS=1;NT=ref;QSS_NT=48;TQSS_NT...</td>\n",
       "      <td>DP:FDP:SDP:SUBDP:AU:CU:GU:TU</td>\n",
       "      <td>163:0:0:0:3,3:4,4:3,3:153,154</td>\n",
       "      <td>200:2:0:0:4,5:8,8:9,9:177,178</td>\n",
       "      <td>scf_1</td>\n",
       "      <td>MSU_v1</td>\n",
       "      <td>CDS</td>\n",
       "      <td>167216</td>\n",
       "      <td>168580</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"N_sal.S000450.1\"; gene_id \"N_sa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>scf_1</td>\n",
       "      <td>860759</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>SOMATIC;QSS=91;TQSS=1;NT=ref;QSS_NT=91;TQSS_NT...</td>\n",
       "      <td>DP:FDP:SDP:SUBDP:AU:CU:GU:TU</td>\n",
       "      <td>150:0:0:0:3,3:147,148:0,0:0,0</td>\n",
       "      <td>231:1:0:0:52,53:178,181:0,0:0,0</td>\n",
       "      <td>scf_1</td>\n",
       "      <td>MSU_v1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>860486</td>\n",
       "      <td>862164</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"N_sal.S002930.1\"; gene_id \"N_sa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>scf_1</td>\n",
       "      <td>947966</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>SOMATIC;QSS=62;TQSS=1;NT=ref;QSS_NT=62;TQSS_NT...</td>\n",
       "      <td>DP:FDP:SDP:SUBDP:AU:CU:GU:TU</td>\n",
       "      <td>131:1:0:0:1,1:1,1:128,130:0,0</td>\n",
       "      <td>154:1:0:0:1,2:14,14:135,137:3,3</td>\n",
       "      <td>scf_1</td>\n",
       "      <td>MSU_v1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>945949</td>\n",
       "      <td>952185</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"N_sal.S003280.1\"; gene_id \"N_sa...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0       1  2  3  4  5     6   \\\n",
       "0  scf_1  167764  .  T  G  .  PASS   \n",
       "1  scf_1  167764  .  T  G  .  PASS   \n",
       "2  scf_1  167764  .  T  G  .  PASS   \n",
       "3  scf_1  860759  .  C  A  .  PASS   \n",
       "4  scf_1  947966  .  G  C  .  PASS   \n",
       "\n",
       "                                                  7   \\\n",
       "0  SOMATIC;QSS=48;TQSS=1;NT=ref;QSS_NT=48;TQSS_NT...   \n",
       "1  SOMATIC;QSS=48;TQSS=1;NT=ref;QSS_NT=48;TQSS_NT...   \n",
       "2  SOMATIC;QSS=48;TQSS=1;NT=ref;QSS_NT=48;TQSS_NT...   \n",
       "3  SOMATIC;QSS=91;TQSS=1;NT=ref;QSS_NT=91;TQSS_NT...   \n",
       "4  SOMATIC;QSS=62;TQSS=1;NT=ref;QSS_NT=62;TQSS_NT...   \n",
       "\n",
       "                             8                              9   \\\n",
       "0  DP:FDP:SDP:SUBDP:AU:CU:GU:TU  163:0:0:0:3,3:4,4:3,3:153,154   \n",
       "1  DP:FDP:SDP:SUBDP:AU:CU:GU:TU  163:0:0:0:3,3:4,4:3,3:153,154   \n",
       "2  DP:FDP:SDP:SUBDP:AU:CU:GU:TU  163:0:0:0:3,3:4,4:3,3:153,154   \n",
       "3  DP:FDP:SDP:SUBDP:AU:CU:GU:TU  150:0:0:0:3,3:147,148:0,0:0,0   \n",
       "4  DP:FDP:SDP:SUBDP:AU:CU:GU:TU  131:1:0:0:1,1:1,1:128,130:0,0   \n",
       "\n",
       "                                10     11      12          13      14      15  \\\n",
       "0    200:2:0:0:4,5:8,8:9,9:177,178  scf_1  MSU_v1  transcript  167216  168580   \n",
       "1    200:2:0:0:4,5:8,8:9,9:177,178  scf_1  MSU_v1        exon  167216  168580   \n",
       "2    200:2:0:0:4,5:8,8:9,9:177,178  scf_1  MSU_v1         CDS  167216  168580   \n",
       "3  231:1:0:0:52,53:178,181:0,0:0,0  scf_1  MSU_v1  transcript  860486  862164   \n",
       "4  154:1:0:0:1,2:14,14:135,137:3,3  scf_1  MSU_v1  transcript  945949  952185   \n",
       "\n",
       "  16 17 18                                                 19  \n",
       "0  .  -  .  transcript_id \"N_sal.S000450.1\"; gene_id \"N_sa...  \n",
       "1  .  -  .  transcript_id \"N_sal.S000450.1\"; gene_id \"N_sa...  \n",
       "2  .  -  0  transcript_id \"N_sal.S000450.1\"; gene_id \"N_sa...  \n",
       "3  .  +  .  transcript_id \"N_sal.S002930.1\"; gene_id \"N_sa...  \n",
       "4  .  -  .  transcript_id \"N_sal.S003280.1\"; gene_id \"N_sa...  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_intersects_0M_R1 = pd.read_csv('../data/nanno_sarek/variant_calling/strelka/0M_R1_vs_537_control_R2/0M_R1_vs_537_control_R2_FILTERED_SNVs_overlap_with_annotations.txt', sep='\\t', header=None)\n",
    "gene_intersects_0M_R1.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74835696-9658-4acb-8f23-501a51df6479",
   "metadata": {},
   "source": [
    "Columns 0-10 are the SNVs data, while 11-19 are from the genome annotation. What we really want is the gene name from the last column for each SNV:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e0d521a1-b23e-4e4f-9526-a39bc88d8b4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_intersects_0M_R1['gene_name'] = gene_intersects_0M_R1[19].str.split('\"').str[1]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd086b2f-90f0-40fe-a431-035af788d7ff",
   "metadata": {},
   "source": [
    "For now, we'll only keep the CDS objects:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "68802761-6b2d-429a-ac72-b1a787d98d12",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_intersects_0M_R1 = gene_intersects_0M_R1[gene_intersects_0M_R1[13] == 'CDS'].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73bb953a-45be-4cac-a896-798c848f8266",
   "metadata": {},
   "source": [
    "How many SNV-affected genes do we have in the non-EMS mutant?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "b4806472-afbc-44a3-b75c-ecacf71b8e78",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "40"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_intersects_0M_R1.groupby(by='gene_name').count().shape[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c3d925e-a98a-4463-88f0-592553716005",
   "metadata": {},
   "source": [
    "Only 40! Now let's see how many of those genes overlap with SNV-affected genes in the EMS mutant; we'll do the command line step for BEDtools intersect again with that file:\n",
    "\n",
    "```\n",
    "bedtools intersect -wa -wb -a ../nanno_sarek/variant_calling/strelka/0_1M_R2_vs_537_control_R2/0_1M_R2_vs_537_control_R2_FILTERED.strelka.somatic_snvs.vcf.gz -b n_salina.working_models.gtf > ../nanno_sarek/variant_calling/strelka/0_1M_R2_vs_537_control_R2/0_1M_R2_vs_537_control_R2_FILTERED_SNVs_overlap_with_annotations.txt\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "59ac6156-6320-4daf-b27c-41e06845360c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>17</th>\n",
       "      <th>18</th>\n",
       "      <th>19</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>scf_1</td>\n",
       "      <td>101592</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>SOMATIC;QSS=97;TQSS=2;NT=ref;QSS_NT=97;TQSS_NT...</td>\n",
       "      <td>DP:FDP:SDP:SUBDP:AU:CU:GU:TU</td>\n",
       "      <td>111:4:0:0:0,0:107,115:0,2:0,0</td>\n",
       "      <td>151:8:0:0:0,1:125,137:0,2:18,18</td>\n",
       "      <td>scf_1</td>\n",
       "      <td>MSU_v1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>101442</td>\n",
       "      <td>104482</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"N_sal.S000260.1\"; gene_id \"N_sa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>scf_1</td>\n",
       "      <td>101592</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>SOMATIC;QSS=97;TQSS=2;NT=ref;QSS_NT=97;TQSS_NT...</td>\n",
       "      <td>DP:FDP:SDP:SUBDP:AU:CU:GU:TU</td>\n",
       "      <td>111:4:0:0:0,0:107,115:0,2:0,0</td>\n",
       "      <td>151:8:0:0:0,1:125,137:0,2:18,18</td>\n",
       "      <td>scf_1</td>\n",
       "      <td>MSU_v1</td>\n",
       "      <td>exon</td>\n",
       "      <td>101442</td>\n",
       "      <td>102074</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"N_sal.S000260.1\"; gene_id \"N_sa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>scf_1</td>\n",
       "      <td>170428</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>SOMATIC;QSS=74;TQSS=1;NT=ref;QSS_NT=74;TQSS_NT...</td>\n",
       "      <td>DP:FDP:SDP:SUBDP:AU:CU:GU:TU</td>\n",
       "      <td>91:8:0:0:0,0:83,93:0,0:0,0</td>\n",
       "      <td>121:18:0:0:0,0:96,114:0,0:7,7</td>\n",
       "      <td>scf_1</td>\n",
       "      <td>MSU_v1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>169388</td>\n",
       "      <td>171263</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"N_sal.S000460.1\"; gene_id \"N_sa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>scf_1</td>\n",
       "      <td>286810</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>SOMATIC;QSS=44;TQSS=2;NT=ref;QSS_NT=44;TQSS_NT...</td>\n",
       "      <td>DP:FDP:SDP:SUBDP:AU:CU:GU:TU</td>\n",
       "      <td>201:0:0:0:1,1:0,0:200,200:0,0</td>\n",
       "      <td>205:1:0:0:5,5:0,0:199,201:0,0</td>\n",
       "      <td>scf_1</td>\n",
       "      <td>MSU_v1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>286705</td>\n",
       "      <td>287415</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"N_sal.S000920.1\"; gene_id \"N_sa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>scf_1</td>\n",
       "      <td>286810</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>SOMATIC;QSS=44;TQSS=2;NT=ref;QSS_NT=44;TQSS_NT...</td>\n",
       "      <td>DP:FDP:SDP:SUBDP:AU:CU:GU:TU</td>\n",
       "      <td>201:0:0:0:1,1:0,0:200,200:0,0</td>\n",
       "      <td>205:1:0:0:5,5:0,0:199,201:0,0</td>\n",
       "      <td>scf_1</td>\n",
       "      <td>MSU_v1</td>\n",
       "      <td>exon</td>\n",
       "      <td>286705</td>\n",
       "      <td>287415</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"N_sal.S000920.1\"; gene_id \"N_sa...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0       1  2  3  4  5     6   \\\n",
       "0  scf_1  101592  .  C  T  .  PASS   \n",
       "1  scf_1  101592  .  C  T  .  PASS   \n",
       "2  scf_1  170428  .  C  T  .  PASS   \n",
       "3  scf_1  286810  .  G  A  .  PASS   \n",
       "4  scf_1  286810  .  G  A  .  PASS   \n",
       "\n",
       "                                                  7   \\\n",
       "0  SOMATIC;QSS=97;TQSS=2;NT=ref;QSS_NT=97;TQSS_NT...   \n",
       "1  SOMATIC;QSS=97;TQSS=2;NT=ref;QSS_NT=97;TQSS_NT...   \n",
       "2  SOMATIC;QSS=74;TQSS=1;NT=ref;QSS_NT=74;TQSS_NT...   \n",
       "3  SOMATIC;QSS=44;TQSS=2;NT=ref;QSS_NT=44;TQSS_NT...   \n",
       "4  SOMATIC;QSS=44;TQSS=2;NT=ref;QSS_NT=44;TQSS_NT...   \n",
       "\n",
       "                             8                              9   \\\n",
       "0  DP:FDP:SDP:SUBDP:AU:CU:GU:TU  111:4:0:0:0,0:107,115:0,2:0,0   \n",
       "1  DP:FDP:SDP:SUBDP:AU:CU:GU:TU  111:4:0:0:0,0:107,115:0,2:0,0   \n",
       "2  DP:FDP:SDP:SUBDP:AU:CU:GU:TU     91:8:0:0:0,0:83,93:0,0:0,0   \n",
       "3  DP:FDP:SDP:SUBDP:AU:CU:GU:TU  201:0:0:0:1,1:0,0:200,200:0,0   \n",
       "4  DP:FDP:SDP:SUBDP:AU:CU:GU:TU  201:0:0:0:1,1:0,0:200,200:0,0   \n",
       "\n",
       "                                10     11      12          13      14      15  \\\n",
       "0  151:8:0:0:0,1:125,137:0,2:18,18  scf_1  MSU_v1  transcript  101442  104482   \n",
       "1  151:8:0:0:0,1:125,137:0,2:18,18  scf_1  MSU_v1        exon  101442  102074   \n",
       "2    121:18:0:0:0,0:96,114:0,0:7,7  scf_1  MSU_v1  transcript  169388  171263   \n",
       "3    205:1:0:0:5,5:0,0:199,201:0,0  scf_1  MSU_v1  transcript  286705  287415   \n",
       "4    205:1:0:0:5,5:0,0:199,201:0,0  scf_1  MSU_v1        exon  286705  287415   \n",
       "\n",
       "  16 17 18                                                 19  \n",
       "0  .  +  .  transcript_id \"N_sal.S000260.1\"; gene_id \"N_sa...  \n",
       "1  .  +  .  transcript_id \"N_sal.S000260.1\"; gene_id \"N_sa...  \n",
       "2  .  +  .  transcript_id \"N_sal.S000460.1\"; gene_id \"N_sa...  \n",
       "3  .  +  .  transcript_id \"N_sal.S000920.1\"; gene_id \"N_sa...  \n",
       "4  .  +  .  transcript_id \"N_sal.S000920.1\"; gene_id \"N_sa...  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_intersects_0_1M_R2 = pd.read_csv('../data/nanno_sarek/variant_calling/strelka/0_1M_R2_vs_537_control_R2/0_1M_R2_vs_537_control_R2_FILTERED_SNVs_overlap_with_annotations.txt', sep='\\t', header=None)\n",
    "gene_intersects_0_1M_R2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "3561a41b-3b85-48b9-b149-218c54370027",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_intersects_0_1M_R2['gene_name'] = gene_intersects_0_1M_R2[19].str.split('\"').str[1]\n",
    "gene_intersects_0_1M_R2 = gene_intersects_0_1M_R2[gene_intersects_0_1M_R2[13] == 'CDS'].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "417ba4bd-75de-4183-ba63-54781083c018",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "345"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_intersects_0_1M_R2.groupby(by='gene_name').count().shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "176417d3-dcd0-4358-a225-2bb5681dc87c",
   "metadata": {},
   "outputs": [],
   "source": [
    "ems_genes = gene_intersects_0_1M_R2.groupby(by='gene_name').count().index.tolist()\n",
    "normal_genes = gene_intersects_0M_R1.groupby(by='gene_name').count().index.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "0f8308df-244a-4031-b63e-9789f3308886",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'N_sal.S003280.1', 'N_sal.S084590.1', 'N_sal.S088740.1'}"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlapping_snps = set(ems_genes).intersection(set(normal_genes))\n",
    "overlapping_snps"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c039049-dec8-4ffa-8834-86fe402a6bc5",
   "metadata": {},
   "source": [
    "Only 3 genes overlap between the two conditions!\n",
    "\n",
    "Now let's look at which indels appear in coding regions:\n",
    "\n",
    "```\n",
    "bedtools intersect -wa -wb -a ../nanno_sarek/variant_calling/strelka/0M_R1_vs_537_control_R2/0M_R1_vs_537_control_R2_FILTERED.strelka.somatic_indels.vcf.gz -b n_salina.working_models.gtf > ../nanno_sarek/variant_calling/strelka/0M_R1_vs_537_control_R2/0M_R1_vs_537_control_R2_FILTERED_indels_overlap_with_annotations.txt\n",
    "\n",
    "bedtools intersect -wa -wb -a ../nanno_sarek/variant_calling/strelka/0_1M_R2_vs_537_control_R2/0_1M_R2_vs_537_control_R2_FILTERED.strelka.somatic_indels.vcf.gz -b n_salina.working_models.gtf > ../nanno_sarek/variant_calling/strelka/0_1M_R2_vs_537_control_R2/0_1M_R2_vs_537_control_R2_FILTERED_indels_overlap_with_annotations.txt\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "aa4327cd-68ed-4ffe-86be-754253f7036e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>17</th>\n",
       "      <th>18</th>\n",
       "      <th>19</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>scf_2</td>\n",
       "      <td>453749</td>\n",
       "      <td>.</td>\n",
       "      <td>AC</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>SOMATIC;QSI=66;TQSI=1;NT=ref;QSI_NT=66;TQSI_NT...</td>\n",
       "      <td>DP:DP2:TAR:TIR:TOR:DP50:FDP50:SUBDP50:BCN50</td>\n",
       "      <td>118:118:103,116:3,3:13,1:118.96:7.27:0:0.06</td>\n",
       "      <td>121:121:72,83:23,26:25,13:116.66:6.6:0:0.05</td>\n",
       "      <td>scf_2</td>\n",
       "      <td>MSU_v1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>452792</td>\n",
       "      <td>455413</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"N_sal.S016490.1\"; gene_id \"N_sa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>scf_11</td>\n",
       "      <td>803969</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>TCGACGG</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>SOMATIC;QSI=109;TQSI=1;NT=ref;QSI_NT=109;TQSI_...</td>\n",
       "      <td>DP:DP2:TAR:TIR:TOR:DP50:FDP50:SUBDP50:BCN50</td>\n",
       "      <td>125:125:119,119:0,0:4,4:137.93:0.01:0:0</td>\n",
       "      <td>192:192:142,143:30,30:18,18:181.19:0.56:0:0</td>\n",
       "      <td>scf_11</td>\n",
       "      <td>MSU_v1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>803966</td>\n",
       "      <td>806674</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"N_sal.S061470.1\"; gene_id \"N_sa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>scf_11</td>\n",
       "      <td>803969</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>TCGACGG</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>SOMATIC;QSI=109;TQSI=1;NT=ref;QSI_NT=109;TQSI_...</td>\n",
       "      <td>DP:DP2:TAR:TIR:TOR:DP50:FDP50:SUBDP50:BCN50</td>\n",
       "      <td>125:125:119,119:0,0:4,4:137.93:0.01:0:0</td>\n",
       "      <td>192:192:142,143:30,30:18,18:181.19:0.56:0:0</td>\n",
       "      <td>scf_11</td>\n",
       "      <td>MSU_v1</td>\n",
       "      <td>exon</td>\n",
       "      <td>803966</td>\n",
       "      <td>804529</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"N_sal.S061470.1\"; gene_id \"N_sa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>scf_11</td>\n",
       "      <td>803969</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>TCGACGG</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>SOMATIC;QSI=109;TQSI=1;NT=ref;QSI_NT=109;TQSI_...</td>\n",
       "      <td>DP:DP2:TAR:TIR:TOR:DP50:FDP50:SUBDP50:BCN50</td>\n",
       "      <td>125:125:119,119:0,0:4,4:137.93:0.01:0:0</td>\n",
       "      <td>192:192:142,143:30,30:18,18:181.19:0.56:0:0</td>\n",
       "      <td>scf_11</td>\n",
       "      <td>MSU_v1</td>\n",
       "      <td>CDS</td>\n",
       "      <td>803966</td>\n",
       "      <td>804529</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"N_sal.S061470.1\"; gene_id \"N_sa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>scf_17</td>\n",
       "      <td>66764</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>CCCTT</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>SOMATIC;QSI=95;TQSI=2;NT=ref;QSI_NT=95;TQSI_NT...</td>\n",
       "      <td>DP:DP2:TAR:TIR:TOR:DP50:FDP50:SUBDP50:BCN50</td>\n",
       "      <td>98:98:87,91:2,5:14,8:95.48:11.66:0:0.12</td>\n",
       "      <td>82:82:50,52:26,28:10,7:77.84:7.28:0:0.09</td>\n",
       "      <td>scf_17</td>\n",
       "      <td>MSU_v1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>63757</td>\n",
       "      <td>70998</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"N_sal.S064940.1\"; gene_id \"N_sa...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       0       1  2   3        4  5     6   \\\n",
       "0   scf_2  453749  .  AC        A  .  PASS   \n",
       "1  scf_11  803969  .   T  TCGACGG  .  PASS   \n",
       "2  scf_11  803969  .   T  TCGACGG  .  PASS   \n",
       "3  scf_11  803969  .   T  TCGACGG  .  PASS   \n",
       "4  scf_17   66764  .   C    CCCTT  .  PASS   \n",
       "\n",
       "                                                  7   \\\n",
       "0  SOMATIC;QSI=66;TQSI=1;NT=ref;QSI_NT=66;TQSI_NT...   \n",
       "1  SOMATIC;QSI=109;TQSI=1;NT=ref;QSI_NT=109;TQSI_...   \n",
       "2  SOMATIC;QSI=109;TQSI=1;NT=ref;QSI_NT=109;TQSI_...   \n",
       "3  SOMATIC;QSI=109;TQSI=1;NT=ref;QSI_NT=109;TQSI_...   \n",
       "4  SOMATIC;QSI=95;TQSI=2;NT=ref;QSI_NT=95;TQSI_NT...   \n",
       "\n",
       "                                            8   \\\n",
       "0  DP:DP2:TAR:TIR:TOR:DP50:FDP50:SUBDP50:BCN50   \n",
       "1  DP:DP2:TAR:TIR:TOR:DP50:FDP50:SUBDP50:BCN50   \n",
       "2  DP:DP2:TAR:TIR:TOR:DP50:FDP50:SUBDP50:BCN50   \n",
       "3  DP:DP2:TAR:TIR:TOR:DP50:FDP50:SUBDP50:BCN50   \n",
       "4  DP:DP2:TAR:TIR:TOR:DP50:FDP50:SUBDP50:BCN50   \n",
       "\n",
       "                                            9   \\\n",
       "0  118:118:103,116:3,3:13,1:118.96:7.27:0:0.06   \n",
       "1      125:125:119,119:0,0:4,4:137.93:0.01:0:0   \n",
       "2      125:125:119,119:0,0:4,4:137.93:0.01:0:0   \n",
       "3      125:125:119,119:0,0:4,4:137.93:0.01:0:0   \n",
       "4      98:98:87,91:2,5:14,8:95.48:11.66:0:0.12   \n",
       "\n",
       "                                            10      11      12          13  \\\n",
       "0  121:121:72,83:23,26:25,13:116.66:6.6:0:0.05   scf_2  MSU_v1  transcript   \n",
       "1  192:192:142,143:30,30:18,18:181.19:0.56:0:0  scf_11  MSU_v1  transcript   \n",
       "2  192:192:142,143:30,30:18,18:181.19:0.56:0:0  scf_11  MSU_v1        exon   \n",
       "3  192:192:142,143:30,30:18,18:181.19:0.56:0:0  scf_11  MSU_v1         CDS   \n",
       "4     82:82:50,52:26,28:10,7:77.84:7.28:0:0.09  scf_17  MSU_v1  transcript   \n",
       "\n",
       "       14      15 16 17 18                                                 19  \n",
       "0  452792  455413  .  -  .  transcript_id \"N_sal.S016490.1\"; gene_id \"N_sa...  \n",
       "1  803966  806674  .  -  .  transcript_id \"N_sal.S061470.1\"; gene_id \"N_sa...  \n",
       "2  803966  804529  .  -  .  transcript_id \"N_sal.S061470.1\"; gene_id \"N_sa...  \n",
       "3  803966  804529  .  -  0  transcript_id \"N_sal.S061470.1\"; gene_id \"N_sa...  \n",
       "4   63757   70998  .  +  .  transcript_id \"N_sal.S064940.1\"; gene_id \"N_sa...  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "indel_intersects_0M_R1 = pd.read_csv('../data/nanno_sarek/variant_calling/strelka/0M_R1_vs_537_control_R2/0M_R1_vs_537_control_R2_FILTERED_indels_overlap_with_annotations.txt', sep='\\t', header=None)\n",
    "indel_intersects_0M_R1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d385f6b8-a224-4507-9b6b-2aa6d3aaa0e7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "indel_intersects_0M_R1['gene_name'] = indel_intersects_0M_R1[19].str.split('\"').str[1]\n",
    "indel_intersects_0M_R1 = indel_intersects_0M_R1[indel_intersects_0M_R1[13] == 'CDS'].reset_index(drop=True)\n",
    "indel_intersects_0M_R1.groupby(by='gene_name').count().shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2e439382-8eeb-4062-abac-ae33e9e03622",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>17</th>\n",
       "      <th>18</th>\n",
       "      <th>19</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>scf_14</td>\n",
       "      <td>627792</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>CCCCTCTCT</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>SOMATIC;QSI=153;TQSI=1;NT=ref;QSI_NT=154;TQSI_...</td>\n",
       "      <td>DP:DP2:TAR:TIR:TOR:DP50:FDP50:SUBDP50:BCN50</td>\n",
       "      <td>109:109:107,107:0,0:2,2:108.73:5.09:0:0.05</td>\n",
       "      <td>96:96:62,65:31,32:3,3:93.88:5.66:0:0.05</td>\n",
       "      <td>scf_14</td>\n",
       "      <td>MSU_v1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>625605</td>\n",
       "      <td>629168</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"N_sal.S080730.1\"; gene_id \"N_sa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>scf_16</td>\n",
       "      <td>149718</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>AAT</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>SOMATIC;QSI=48;TQSI=1;NT=ref;QSI_NT=48;TQSI_NT...</td>\n",
       "      <td>DP:DP2:TAR:TIR:TOR:DP50:FDP50:SUBDP50:BCN50</td>\n",
       "      <td>30:30:24,24:0,0:4,5:39.61:0.27:0:0</td>\n",
       "      <td>28:28:22,23:4,4:2,2:36.04:0.47:0:0</td>\n",
       "      <td>scf_16</td>\n",
       "      <td>MSU_v1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>147799</td>\n",
       "      <td>151161</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"N_sal.S073150.2\"; gene_id \"N_sa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>scf_16</td>\n",
       "      <td>149718</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>AAT</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>SOMATIC;QSI=48;TQSI=1;NT=ref;QSI_NT=48;TQSI_NT...</td>\n",
       "      <td>DP:DP2:TAR:TIR:TOR:DP50:FDP50:SUBDP50:BCN50</td>\n",
       "      <td>30:30:24,24:0,0:4,5:39.61:0.27:0:0</td>\n",
       "      <td>28:28:22,23:4,4:2,2:36.04:0.47:0:0</td>\n",
       "      <td>scf_16</td>\n",
       "      <td>MSU_v1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>147799</td>\n",
       "      <td>151367</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"N_sal.S073150.1\"; gene_id \"N_sa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>scf_16</td>\n",
       "      <td>149718</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>AAT</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>SOMATIC;QSI=48;TQSI=1;NT=ref;QSI_NT=48;TQSI_NT...</td>\n",
       "      <td>DP:DP2:TAR:TIR:TOR:DP50:FDP50:SUBDP50:BCN50</td>\n",
       "      <td>30:30:24,24:0,0:4,5:39.61:0.27:0:0</td>\n",
       "      <td>28:28:22,23:4,4:2,2:36.04:0.47:0:0</td>\n",
       "      <td>scf_16</td>\n",
       "      <td>MSU_v1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>147813</td>\n",
       "      <td>151295</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"N_sal.S073150.3\"; gene_id \"N_sa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>scf_20</td>\n",
       "      <td>158358</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>AG</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>SOMATIC;QSI=38;TQSI=1;NT=ref;QSI_NT=38;TQSI_NT...</td>\n",
       "      <td>DP:DP2:TAR:TIR:TOR:DP50:FDP50:SUBDP50:BCN50</td>\n",
       "      <td>36:36:35,39:0,0:1,1:34.41:2.26:0:0.06</td>\n",
       "      <td>57:57:47,48:6,6:4,6:51.67:1.93:0:0.02</td>\n",
       "      <td>scf_20</td>\n",
       "      <td>MSU_v1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>152063</td>\n",
       "      <td>162621</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"N_sal.S050720.1\"; gene_id \"N_sa...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       0       1  2  3          4  5     6   \\\n",
       "0  scf_14  627792  .  C  CCCCTCTCT  .  PASS   \n",
       "1  scf_16  149718  .  A        AAT  .  PASS   \n",
       "2  scf_16  149718  .  A        AAT  .  PASS   \n",
       "3  scf_16  149718  .  A        AAT  .  PASS   \n",
       "4  scf_20  158358  .  A         AG  .  PASS   \n",
       "\n",
       "                                                  7   \\\n",
       "0  SOMATIC;QSI=153;TQSI=1;NT=ref;QSI_NT=154;TQSI_...   \n",
       "1  SOMATIC;QSI=48;TQSI=1;NT=ref;QSI_NT=48;TQSI_NT...   \n",
       "2  SOMATIC;QSI=48;TQSI=1;NT=ref;QSI_NT=48;TQSI_NT...   \n",
       "3  SOMATIC;QSI=48;TQSI=1;NT=ref;QSI_NT=48;TQSI_NT...   \n",
       "4  SOMATIC;QSI=38;TQSI=1;NT=ref;QSI_NT=38;TQSI_NT...   \n",
       "\n",
       "                                            8   \\\n",
       "0  DP:DP2:TAR:TIR:TOR:DP50:FDP50:SUBDP50:BCN50   \n",
       "1  DP:DP2:TAR:TIR:TOR:DP50:FDP50:SUBDP50:BCN50   \n",
       "2  DP:DP2:TAR:TIR:TOR:DP50:FDP50:SUBDP50:BCN50   \n",
       "3  DP:DP2:TAR:TIR:TOR:DP50:FDP50:SUBDP50:BCN50   \n",
       "4  DP:DP2:TAR:TIR:TOR:DP50:FDP50:SUBDP50:BCN50   \n",
       "\n",
       "                                           9   \\\n",
       "0  109:109:107,107:0,0:2,2:108.73:5.09:0:0.05   \n",
       "1          30:30:24,24:0,0:4,5:39.61:0.27:0:0   \n",
       "2          30:30:24,24:0,0:4,5:39.61:0.27:0:0   \n",
       "3          30:30:24,24:0,0:4,5:39.61:0.27:0:0   \n",
       "4       36:36:35,39:0,0:1,1:34.41:2.26:0:0.06   \n",
       "\n",
       "                                        10      11      12          13  \\\n",
       "0  96:96:62,65:31,32:3,3:93.88:5.66:0:0.05  scf_14  MSU_v1  transcript   \n",
       "1       28:28:22,23:4,4:2,2:36.04:0.47:0:0  scf_16  MSU_v1  transcript   \n",
       "2       28:28:22,23:4,4:2,2:36.04:0.47:0:0  scf_16  MSU_v1  transcript   \n",
       "3       28:28:22,23:4,4:2,2:36.04:0.47:0:0  scf_16  MSU_v1  transcript   \n",
       "4    57:57:47,48:6,6:4,6:51.67:1.93:0:0.02  scf_20  MSU_v1  transcript   \n",
       "\n",
       "       14      15 16 17 18                                                 19  \n",
       "0  625605  629168  .  -  .  transcript_id \"N_sal.S080730.1\"; gene_id \"N_sa...  \n",
       "1  147799  151161  .  -  .  transcript_id \"N_sal.S073150.2\"; gene_id \"N_sa...  \n",
       "2  147799  151367  .  -  .  transcript_id \"N_sal.S073150.1\"; gene_id \"N_sa...  \n",
       "3  147813  151295  .  -  .  transcript_id \"N_sal.S073150.3\"; gene_id \"N_sa...  \n",
       "4  152063  162621  .  -  .  transcript_id \"N_sal.S050720.1\"; gene_id \"N_sa...  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "indel_intersects_0_1M_R2 = pd.read_csv('../data/nanno_sarek/variant_calling/strelka/0_1M_R2_vs_537_control_R2/0_1M_R2_vs_537_control_R2_FILTERED_indels_overlap_with_annotations.txt', sep='\\t', header=None)\n",
    "indel_intersects_0_1M_R2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3b8e6a98-802f-4d62-9318-90dfaa6cf5bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "indel_intersects_0_1M_R2['gene_name'] = indel_intersects_0_1M_R2[19].str.split('\"').str[1]\n",
    "indel_intersects_0_1M_R2 = indel_intersects_0_1M_R2[indel_intersects_0_1M_R2[13] == 'CDS'].reset_index(drop=True)\n",
    "indel_intersects_0_1M_R2.groupby(by='gene_name').count().shape[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0cce22ce-2b68-4aaf-b80e-d2dee6c84bec",
   "metadata": {},
   "source": [
    "Only 3 indels overlap coding regions -- what genes are they?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "bd8e175a-5e98-4aa3-addc-18a4a5eebc0a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(['N_sal.S061470.1', 'N_sal.S084590.1'], ['N_sal.S034610.1'])"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "indel_genes_0M_R1 = indel_intersects_0M_R1.groupby(by='gene_name').count().index.tolist()\n",
    "indel_genes_0_1M_R2 = indel_intersects_0_1M_R2.groupby(by='gene_name').count().index.tolist()\n",
    "indel_genes_0M_R1, indel_genes_0_1M_R2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a0c0090-b93d-4bbc-98ad-f07202361a65",
   "metadata": {},
   "source": [
    "None of them overlap between the two mutation conditions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "2ae112ab-2d34-413a-b8a4-395b4ae2a93e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlapping_indels = set(indel_genes_0M_R1).intersection(set(indel_genes_0_1M_R2))\n",
    "overlapping_indels"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "834219e4-1dea-4de2-b1f3-0434f4d2c3ee",
   "metadata": {},
   "source": [
    "Do any of them appear in genes that also have SNPs?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "81094967-138f-4a04-b04a-510b89ea8f97",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'N_sal.S084590.1'}"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(indel_genes_0M_R1).intersection(set(normal_genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "72f16084-b0bb-410f-8dc7-b14ec780a5cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(indel_genes_0_1M_R2).intersection(set(ems_genes))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "omics",
   "language": "python",
   "name": "omics"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
